lib/checks/aria/required-children.js

```diff
@@ -106,49 +135,85 @@ function missingRequiredChildren(node, childRoles, all, role) {
 	return null;
 }
 
-function hasDecendantWithRole(node) {
-	return (
-		node.children &&
-		node.children.some(child => {
-			const role = axe.commons.aria.getRole(child);
-			return (
-				!['presentation', 'none', null].includes(role) ||
-				hasDecendantWithRole(child)
-			);
-		})
-	);
-}
-
-const role = node.getAttribute('role');
-const required = requiredOwned(role);
+/**
+ * Helper to check if a given node owns an element with a given role
+ * @param {HTMLElement} node node
+ * @param {Object} virtualTree virtual node
+ * @param {String} role role
+ * @param {Boolean} ariaOwned
+ * @returns {Boolean}
+ */
+function owns(node, virtualTree, role, ariaOwned) {
+	if (node === null) {
+		return false;
+	}
+	const implicit = implicitNodes(role);
+	let selector = ['[role="' + role + '"]'];
 
-if (!required) {
-	return true;
-}
+	if (implicit) {
+		selector = selector.concat(
+			implicit.map(implicitSelector => implicitSelector + ':not([role])')
+		);
+	}
 
-let all = false;
-let childRoles = required.one;
-if (!childRoles) {
-	all = true;
-	childRoles = required.all;
+	selector = selector.join(',');
+	return ariaOwned
+		? matchesSelector(node, selector) ||
+				!!querySelectorAll(virtualTree, selector)[0]
+		: !!querySelectorAll(virtualTree, selector)[0];
 }
 
-const missing = missingRequiredChildren(node, childRoles, all, role);
+/**
+ * Helper to check if a given node is `aria-owns` an element with a given role
+ * @param {HTMLElement[]} nodes nodes
+ * @param {String} role role
+ * @returns {Boolean}
+ */
+function ariaOwns(nodes, role) {
+	for (let index = 0; index < nodes.length; index++) {
+		const node = nodes[index];
+		if (node === null) {
+			continue;
+		}
 
-if (!missing) {
-	return true;
+		const virtualTree = getNodeFromTree(node);
+		if (owns(node, virtualTree, role, true)) {
+			return true;
+		}
+	}
+	return false;
 }
 
-this.data(missing);
+/**
+ * Get role (that is not presentation or none) of descendant
+ * @param {HTMLElement} node node
+ * @returns {String[]}
+ */
+function getDescendantRole(node, ownedEls) {
+	const isOwns = ownedEls && ownedEls.length > 0;
+	const el = isOwns ? ownedEls[0] : node;
+
+	if (!el) {
+		return [];
+	}
 
-// Only review empty nodes when a node is both empty and does not have an aria-owns relationship
-if (
-	reviewEmpty.includes(role) &&
-	!hasContentVirtual(virtualNode, false, true) &&
-	!hasDecendantWithRole(virtualNode) &&
-	idrefs(node, 'aria-owns').length === 0
-) {
-	return undefined;
-} else {
-	return false;
+	const items = isOwns
+		? Array.from(el.children).reduce(
```

Comment:
What exactly is this reduce doing? It looks like it's just pushing the children onto an array with the parent element. If that's the case it would be cleaner to just use concat.

```js
const items = (isOwns ? [el] : []).concat(Array.from(el.children));
```